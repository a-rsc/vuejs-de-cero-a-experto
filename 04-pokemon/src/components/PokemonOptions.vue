<template>
    <div class="options-container">
        <!-- Utilizaremos la api https://pokeapi.co para añadir nombres de diferentes pokemons -->
        <ul>
            <!-- $emit: el símbolo de $ es un caracter normal y corriente. ¿Por qué empieza con el símbolo de $? El símbolo de $ representa las propiedades y eventos propios de Vue, para evitar que un desarrollador los utilice.  -->
            <!-- https://v3.vuejs.org/guide/component-custom-events.html#event-names -->
            <li v-for="pokemon in pokemons" :key="pokemon.id" @click="$emit('selectedPokemon', pokemon.id)">{{ pokemon.name }}</li>
        </ul>
    </div>
</template>

<script>
export default {
    name: 'Options',
    components: {
    },
    props: {
        pokemons: {
            type: Array,
            required: true
        }
    },
    data() {
        return {
        }
    },
    methods: {
        hola() {
            // this.$emit = ...
        }
    }
}
</script>

<style scoped>
.options-container {
    display: flex;
    justify-content: center;
}
ul {
    padding: 0;
    list-style-type: none;
}
li {
    line-height: 32px;

    background-color: white;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
}
li:hover {
    background-color: rgba(0, 0, 0, 0.05);
}
</style>