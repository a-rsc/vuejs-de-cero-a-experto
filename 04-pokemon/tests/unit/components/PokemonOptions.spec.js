import { shallowMount } from '@vue/test-utils'
import PokemonOptions from '@/components/PokemonOptions'

import {pokemons} from '../mocks/pokemons.mock'

// npm run test:unit /PokemonOptions
describe('PokemonOptions Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonOptions, {
            props: {
                // pokemons: pokemons // ECMAScript6
                pokemons
            }
        })
    })

    // la funci칩n it es muy parecida
    test('Debe de hacer match con el snapshot', () => {
        // console.log(wrapper.html()) // vemos lo que se est치 renderizando
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar las 4 opciones correctamente', () => {
        // que existan 4 li
        const liTags = wrapper.findAll('li')
        expect(liTags.length).toBe(4)

        // cada li debe de tener su nombre del pokemon
        expect(liTags[0].text()).toBe('Bulbasaur')
        expect(liTags[1].text()).toBe('Ivysaur')
        expect(liTags[2].text()).toBe('Venusaur')
        expect(liTags[3].text()).toBe('Charmander')
    })

    test('Debe de emitir "selection" con sus respectivos par치metros al hacer click', () => {
        const [li1, li2, li3, li4] = wrapper.findAll('li')

        li1.trigger('click')
        // console.log(wrapper.emitted('selectionPokemon'))
        expect(wrapper.emitted('selectionPokemon').length).toBe([1]) // el evento se emiti칩 una vez
        // expect(wrapper.emitted(selectionPokemon)[0]).toBe(1)
        // [] === [] // false
        // {} === {} // false

        expect(wrapper.emitted(selectionPokemon)[0]).toEqual([1])
    })

})
