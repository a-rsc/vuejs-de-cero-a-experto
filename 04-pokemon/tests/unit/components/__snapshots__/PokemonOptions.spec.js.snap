// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions Component Debe de hacer match con el snapshot 1`] = `
<div class="options-container">
  <!-- Utilizaremos la api https://pokeapi.co para añadir nombres de diferentes pokemons -->
  <ul>
    <!-- $emit: el símbolo de $ es un caracter normal y corriente. ¿Por qué empieza con el símbolo de $? El símbolo de $ representa las propiedades y eventos propios de Vue, para evitar que un desarrollador los utilice.  -->
    <!-- https://v3.vuejs.org/guide/component-custom-events.html#event-names -->
    <li>Bulbasaur</li>
    <li>Ivysaur</li>
    <li>Venusaur</li>
    <li>Charmander</li>
  </ul>
</div>
`;
